feat(lsp): add `vim.lsp.inlay_hint.action()` (#36219)
neovim/neovim: runtime/doc/lsp.txt

                    • `"textEdits"`: insert `textEdits` that comes with the
                      inlay hints.
                    • `"location"`: jump to one of the locations associated
                      with the inlay hints.
                    • `"command"`: execute one of the `lsp.Command`s that
                      comes with the inlay hint.
                    • `"hover"`: if there are some locations associated with
                      the inlay hint, show the hover information of the
                      identifiers at those locations.
                    • `"tooltip"`: show a hover-like window that contains the
                      `tooltip`, available `command`s and `location`s that
                      comes with the inlay hint.
                    • a custom handler with 3 parameters:
                      • `hints`: `lsp.InlayHint[]` a list of inlay hints in
                        the requested range.
                      • `ctx`: `{bufnr: integer, client: vim.lsp.Client}` the
                        buffer number on which the action is taken, and the
                        LSP client that provides `hints`.

>                       • `on_finish`:

Reviewer: ```suggestion
                      • `on_done`:
```

>                       • `on_finish`:

Reviewer: do these parameters (order, name, shape) align with other comparable existing lsp features?

==============================================================================
Lua module: vim.lsp.inlay_hint                                *lsp-inlay_hint*

apply_action({action}, {opts})             *vim.lsp.inlay_hint.apply_action()*

>     For supported LSP servers, apply one of the following actions provided by

Reviewer 2: ```suggestion
    Apply one of the following actions provided by
```
It is already implied that this will use language servers that support the feature.

    inlayhints in the selected range.
    • In |Normal-mode|, the action applies to inlay hints that are adjacent to
      the cursor.
    • In |Visual-mode|, the action applies to inlay hints that are in the
      visually selected range.

    Example usage: >lua
        vim.keymap.set(
          'n',
          'gI',
          function()

>             vim.lsp.inlay_hint.apply_action("textEdits")

Reviewer 2: Nit:
```suggestion
            vim.lsp.inlay_hint.apply_action('textEdits')
```

          end,
          { desc = 'Apply inlay hint edits' }
        )
<

    Parameters: ~
      • {action}  (`vim.lsp.inlay_hint.action`)

                  Possible actions: `"textEdits"`, `"tooltip"`, `"location"`,
                  `"command"` or a custom callback:
                  `fun(hints: lsp.InlayHint[], ctx: vim.lsp.inlay_hint.action.context):integer`,

>                   which accepts the resolved inlayhints in the range and some

Reviewer 2: ```suggestion
                  which accepts the resolved inlay hints in the range and some
```

==============================================================================
Lua module: vim.lsp.inlay_hint                                *lsp-inlay_hint*

                                           *vim.lsp.inlay_hint.apply_action()*
apply_action({action}, {opts}, {callback})
    Apply some actions provided by inlay hints in the selected range.

    Example usage: >lua
        vim.keymap.set(
          { 'n', 'v' },

>           'gI',

Reviewer: > The "action" here isn't the same as code actions. These are attributes of the inlay hints that we can "take actions on" (locations, textedits, tooltips, etc.).

Of course it's not exactly the same, but the concept is the same: a list of actions that actions on a range or nearby position.

> > Is there a way to list the available inlay-hint actions? (Is that just "list the inlay hints"?)
> 
> `vim.lsp.inlay_hint.apply_action("tooltip")` does this ... This'll tell the user

No, I mean a programmatic way to list stuff.

==============================================================================
Lua module: vim.lsp.inlay_hint                                *lsp-inlay_hint*

*vim.lsp.inlay_hint.apply_text_edits.Opts*

    Fields: ~

>       • {post_filtering}  (`boolean?`) Whether to filter the inlay hints to

Reviewer 2: Why should we make this an option? Intuitively I would expect this to only resolve the inlay hints under the cursor.

                          strictly include the ones in the range.

apply_text_edits({opts})               *vim.lsp.inlay_hint.apply_text_edits()*
    For supported LSP servers, apply the `textEdit`s in the inlay hint to the
    buffer.
    • In |Normal-mode|, this function inserts inlay hints that are adjacent to
      the cursor.
    • In |Visual-mode|, this function inserts inlay hints that are in the
      visually selected range.

    Example usage: >lua
        vim.keymap.set("n", "gI", vim.lsp.inlay_hint.apply_text_edits, {})
<

>     For dot-repeat, you can set up the keymap like the following: >lua

Reviewer 2: This feels a bit niche. I would leave it out of the docs.

Author: @ribru17 what do you think about this? if we're dropping this section from the doc, I might as well just drop [this `if` block here](https://github.com/Davidyz/neovim/blob/52fb934668c922c13a579367024cc01bf7885197/runtime/lua/vim/lsp/inlay_hint.lua#L469).

Author: I think I'll just drop built-in dot-repeat, because it's tricky (and maybe too complicated) to get it to play nicely with `opts`. Users can always create a wrapper for the `apply_text_edits` function with the preferred options and use the wrapper with `operatorfunc`.

Reviewer 2: Yeah dot-repeat isn't something that we should worry about in the initial implementation of LSP features.
